Retry failures in couch_mrview_compactor:recompact

jaydoane opened the pull request:
Currently, if there's an error in the spawned couch_index_updater:update
process, the receive block will not match the message, and the calling
process will block forever.

This commit changes the receive pattern to match both successful updates
as well as errors, logging the errors, and retrying the recompact a
configurable number of times. If the configured retry count is exceeded,
it logs an error, and returns the original state.

COUCHDB-3184

jaydoane commented:
@davisp, not sure if this is exactly what you had in mind, but I thought I'd take a stab at it to get the ball rolling

davisp commented:
That's the general idea, though there was a bit I wanted to add using the update/3 so that we know if the updater is making progress between failures. I typed up a quick thing but haven't tried compiling it or anything crazy like that.

```diff
diff --git a/src/couch_mrview_compactor.erl b/src/couch_mrview_compactor.erl
index 9dba094..3271103 100644
--- a/src/couch_mrview_compactor.erl
+++ b/src/couch_mrview_compactor.erl
@@ -148,14 +148,31 @@ compact(State) ->
 
 
 recompact(State) ->
+    recompact(State, recompact_retry_count()).
+
+recompact(State, 0) ->
+    erlang:error(exceeded_recompact_retry_count);
+
+recompact(State, RetryCount) ->
+    Self = self(),
     link(State#mrst.fd),
     {Pid, Ref} = erlang:spawn_monitor(fun() ->
-        couch_index_updater:update(couch_mrview_index, State)
+        couch_index_updater:update(Self, couch_mrview_index, State)
     end),
+    recompact_loop(State, RetryCount).
+
+recompact_loop(State, RetryCount) ->
     receive
+        {'$gen_cast', {new_state, State}} ->
+            % We've made progress so reset RetryCount
+            recompact_loop(State, recompact_retry_count());
         {'DOWN', Ref, _, _, {updated, Pid, State2}} ->
             unlink(State#mrst.fd),
-            {ok, State2}
+            {ok, State2};
+        {'DOWN', Ref, _, _, Reason} ->
+            unlink(State#mrst.fd),
+            couch_log:warning("Error durring recompaction: ~r", [Reason]),
+            recompact(State, RetryCount - 1)
     end.
 
 compact_log(LogBtree, BufferSize, Acc0) ->
@@ -265,3 +282,11 @@ swap_compacted(OldState, NewState) ->
     erlang:demonitor(OldState#mrst.fd_monitor, [flush]),
     
     {ok, NewState#mrst{fd_monitor=Ref}}.
+
+
+recompact_retry_count() ->
+    config:get_integer(
+        "view_compaction",
+        "recompact_retry_count",
+        ?DEFAULT_RECOMPACT_RETRY_COUNT
+    ).
```

davisp commented:
Already see a bug, The $gen_cast message needs to use a different variable name which it passes to recompact_loop so that we don't get a bad match.

jaydoane commented:
@davisp, thanks for the diff. I updated the PR, added `Pid` and `Ref` params to `recompact_loop` function, and fixed the `State` name issue for `'$gen_cast'` case.

One concern is that while we have a test for `couch_mrview:compact/3`, we don't actually test `recompact` in it. Do you think it's sufficient to just unit test `recompact` and `recompact_loop`, or do you think it warrants a full on integration style test such as in `couch_mrview_compact_tests`?

davisp commented:
I'd be fine with just adding unit tests. I don't know how we'd test behavior very well on this with an integration test as the bad cases are all when views get huge, which isn't super amenable to CI tests.

jaydoane commented on src/couch_mrview_compactor.erl:
```diff
@@ -148,16 +149,41 @@ compact(State) ->
+            {ok, State2};
+        {'DOWN', Ref, _, _, Reason} ->
+            unlink(State#mrst.fd),
+            couch_log:warning("Error during recompaction: ~r", [Reason]),
```
shouldn't this be `~p`?

jaydoane commented:
@davisp, I added a couple tests which seem to exercise all the new retry/progress logic.

I'm not familiar with the `~r` control sequence in the `couch_log:warning`, nor am I able to find it in http://erlang.org/doc/man/io.html, but it doesn't seem to cause any problems, either.

davisp commented:
@jaydoane The ~r is a custom control character I added to couch_log to simplify formatting error terms:

https://github.com/apache/couchdb-couch-log/commit/038586b7ab6cfec1b15f44162eddb5723803be21

jaydoane commented:
@davisp, thanks for the review. Can you merge this for me?
